src/hamiltonian.jl

```diff
@@ -0,0 +1,61 @@
+# TODO: add a type for kinetic energy
+
+struct Hamiltonian{M<:AbstractMetric}
+    metric      ::  M
+    _logπ       ::  Function
+    _dlogπdθ    ::  Function
+end
+
+# TODO: implement a helper function for those only _logπ is provided and use AD to provide _dlogπdθ
+
+function _dHdθ(h::Hamiltonian, θ::AbstractVector{T}) where {T<:Real}
+    return -h._dlogπdθ(θ)
+end
+
+function _dHdr(h::Hamiltonian{UnitMetric}, r::AbstractVector{T}) where {T<:Real}
+    return r
+end
+
+function _dHdr(h::Hamiltonian{DiagMetric{T}}, r::AbstractVector{T}) where {T<:Real}
+    return h.metric.M⁻¹ .* r
+end
+
+function _dHdr(h::Hamiltonian{DenseMetric{T}}, r::AbstractVector{T}) where {T<:Real}
+    return h.metric.M⁻¹ * r
+end
+
+function _H(h::Hamiltonian, θ::AbstractVector{T}, r::AbstractVector{T}) where {T<:Real}
+    return _K(h, r, θ) + _V(h, θ)
+end
+
+function _V(h::Hamiltonian, θ::AbstractVector{T}) where {T<:Real}
+    return -h._logπ(θ)
+end
+
+# Kinetic energy
+# NOTE: the general form of K depends on both θ and r
+function _K(h::Hamiltonian{UnitMetric}, r::AbstractVector{T}, ::AbstractVector{T}) where {T<:Real}
+    return sum(abs2, r) / 2
+end
+
+function _K(h::Hamiltonian{DiagMetric{T}}, r::AbstractVector{T}, ::AbstractVector{T}) where {T<:Real}
+    return sum(abs2, r .* sqrt.(h.metric.M⁻¹)) / 2
+end
+
+function _K(h::Hamiltonian{DenseMetric{T}}, r::AbstractVector{T}, ::AbstractVector{T}) where {T<:Real}
+    return p' * h.metric.M⁻¹ * p / 2
```

Comment:
In this case `Metric` would be mutable. Does it affect any efficiency?

Reply:
BTW I'm happy to change it to mutable if this doesn't have any efficiency drawback.

Reply:
No, `Metric` doesn't need to be mutable. `@! h.temp = h.metric.M⁻¹ * p` is syntactic sugar for `mul!(h.temp, h.metric.M⁻¹, p)` which writes the result of the multiplication into `h.temp`.

Reply:
OK I meant to ask about the mutability about `Hamiltonian`. Is that a concern?

Reply:
OK I realised that there should be no issue.

Reply:
@mohamed82008 I've resolved all memory comments from you except this one (they do give me some improvement! thanks again!). For this one, only `DenseEuclideanMetric` need this `temp` variable, which seems to be unnecessary if I put it as a field of `h` as other metric types do not use it. Any idea how I can work around it? Is there a way to do this computation similar to what you suggest via `BroadcastArray`?

Reply:
In general immutable types are more performant than mutable ones, because they are stack allocated and incur no allocations or loading time from the heap memory. I also find immutable types easier to reason about in many cases. 

As for the `temp` problem, linear algebra on normal arrays in Julia is done with the heavily optimized BLAS library which is typically faster than a hand-written loop of the naive approach. For instance, `sum(A[i,j]*b[i]*b[j] for i in 1:size(A, 1), j in 1:size(A, 2))` is mathematically equivalent to `b' * A * b` but the latter is generally faster.  Unfortunately, I can't think of any workaround as BLAS doesn't have a routine for this extremely common operation, aka Galerkin projection. I plan to give this branch a second thorough read next weekend and make a PR to your branch if I see some bugs or obvious improvements. You can think of a way to reuse the temporaries which are not used anymore in the rest of the current iteration. If there is no way, then try to benchmark the pre-allocated and normal versions and see which is faster.

Reply:
I see. Thanks!

Reply:
OK. I realised that `h.temp` in your example is just `h._dHdr`. Maybe I can just reuse that allocation.